### ISSUE TITLE
RFC: State specific Systems and Scenes

### ISSUE THREAD

**Issue by Xaeroxe** [amethyst/amethyst#379]
As a player traverses through menus and game levels each will likely have its own `State` associated with it.

As the game transitions between States it may be desirable for the game programmer to introduce and remove Systems, and to remove Entities when the context in which they are necessary is no longer present.  Simultaneously however, some Systems and Entities will need to be global.  So in order to achieve these goals I propose two additions to the main crate:

1. State specific dispatchers.  Each State has a Dispatcher associated with it in the StateMachine, the State can populate the Dispatcher with Systems in `on_start` and this dispatcher will be executed after the primary global Dispatcher.  This Dispatcher will only be disposed of if the State structure associated with it is also disposed of, either by `Trans::Pop` or `Trans::Switch`

2. State specific `Scenes`.  Each state will have access to a `Scene` structure which it can modify in any of functions on the `State` trait.  A `Scene` is simply a vector of Entities correlated with this `State`.  When the `State` is disposed of, either by `Trans::Pop` or `Trans::Switch` all entities in the Scene will be deleted from the specs::World.

EDIT: In order to keep the `Scene` vector from growing indefinitely it will be iterated through every frame and any entities found that are no longer [alive](https://docs.rs/specs/0.9.5/specs/struct.EntitiesRes.html#method.is_alive) will be `swap_remove`d from the vector.

Unresolved question: Should entities that are part of a Scene be made "inactive" when the State is paused, and if so how would this be implemented?

**Comment by Xaeroxe** [comment]
cc @amethyst/engine-devs This will impact all of us so please review when you have time.

**Comment by jojolepro** [comment]
Unresolved question: Should entities that are part of a Scene be made "inactive" when the State is paused, and if so how would this be implemented?
I'd say yes. How to do it? Remove them from specs on pause, and put them back in on resume. You already talked about a scene. From what I understand, its a buffer that tracks which entities were created by this State. We could make the Systems owned by a State register their new Entities as either owned by the Global game, or by the local/current State.

Also as a reference, android components look really closely to what we are trying to achieve with State. The only big difference that I see is that we also have to sync our data with a multithreaded ecs, which is both very efficient, and very complex. We'll have to innovate quite a bit here ;)
![image](https://user-images.githubusercontent.com/4755185/30675021-08545854-9e4c-11e7-9ebc-8ab282e23708.png)

**Comment by Xaeroxe** [comment]
Restoring an entity after it's been deleted is non-trivial, but it might be possible to carefully extract and store the components related to them.  It would require navigating the type system a bit but it might be possible.

Systems actually can't create new entities, and my idea for States is that when creating new Entities they could either add it to the scene, which will cause the entity to be deleted on removal of the State, or simply not add it to the scene, which would cause the entity to be treated as a global entity.

**Comment by jojolepro** [comment]
> Systems actually can't create new entities. 

I actually needed that to optimize my game. Fps can get as 3 times higher when not adding entities that are offscreen (for 120 entities, where the maximum I ever saw was around 15000).

> Restoring an entity after it's been deleted is non-trivial.

@omni-viral serialization for entities and component <3
Not sure about the cost of it.

> and my idea for States is that when creating new Entities they could either add it to the scene, which will cause the entity to be deleted on removal of the State, or simply not add it to the scene, which would cause the entity to be treated as a global entity.

:+1:

**Comment by Xaeroxe** [comment]
Some `State`s may not want to inactivate their `Scene`'s entities on pause, so perhaps the Scene should also contain the ability to opt out of this behavior.

**Comment by jojolepro** [comment]
Do you have a use case for it?

**Comment by Xaeroxe** [comment]
After thoughts on above comment: The same idea can be implemented without the opt out feature.  Just make the entity a global entity and have the state remove it on exit.  So between that and Keep It Simple Stupid, I no longer believe we need an "opt out" feature for inactivation.

**Comment by Xaeroxe** [comment]
Thoughts on Inactivation implementation:

Most methods of adding and removing an entity from a world will, out of necessity, change the entity's index.  This stinks because it means any stored Entity values won't fetch that Entity now.  Another approach would be to just remove an Entity's components and restore them later, but components can still be added to this entity when it is "Inactive" allowing what may be possible bugs with handling "Inactive" entities to persist.

So another approach is to not make any attempt at preventing users from modifying Inactive entities, but instead to mark the Active ones with an `Active` component.  Under this approach most systems will query for the `Active` component in addition to their regular components.

**Comment by torkleyy** [comment]
> Systems actually can't create new entities,

Yes, they can. Just use `Entities::create`.

**Comment by omni-viral** [comment]
My entity serialization approach is limited to defined at startup set of components

**Comment by Rhuagh** [comment]
You can already access the world in state transition so the second part is already handled. There are really cool ways of deleting entities already, thanks to the efforts of the almighty @torkleyy.
I don't think we need the extra layer of Scene.

For the systems, it gets harder. I like your approach of having a dispatcher per state, and one for the global. You might have global systems that you want to run before and some after the state systems though, and this might differ per state. Then again, you can always move some of the "global" systems into your state dispatcher, so I guess that's fine.

**Comment by Rhuagh** [comment]
> I have a game with many game modes, one of these modes takes 3 rounds, I have a state that keeps track of how many rounds have happened and who won what rounds, this state eventually Pushes the "Game Play" state after it has determined another round should occur. In between rounds I also have entities that should be preserved, such as the announcer character in the middle of the map. I also want this announcer character to go away once my "GameRounds" state is popped from the stack.

In your GamePlay on_end: 

```rust
let entities = (&*world.entities(), &world.read::<SomeComponent>).join().collect();
world.delete_entities(&entities);
```

GameRounds::on_end:

```rust
let entities = (&*world.entities(), &world.read::<GameAnnouncer>).join().collect();
world.delete_entities(&entities);
```

**Comment by Xaeroxe** [comment]
@Rhuagh while the scene abstraction isn't strictly necessary it can be very convenient, especially given most games will probably implement something similar to it anyways.

**Comment by torkleyy** [comment]
@Xaeroxe Our first concern right now should be fixing the existing abstractions and providing the functionality, instead of abstracting to a point where it's either too complicated or too inflexible. So, in general, states should be integrated and yes, we need multiple dispatchers (or at least change the one), but I'm not sure if that's what we should be implementing right now, given that there's no game bigger than pong implemented in Amethyst.

**Comment by Xaeroxe** [comment]
The cool thing about both of these proposals is they're purely additive.  If you simply don't use them then they won't impact your game design at all.

**Comment by Xaeroxe** [comment]
@torkleyy this actually came up because @jojolepro wanted features like this for the game he's working on.

**Comment by Xaeroxe** [comment]
It would also take fairly little time to implement this, maybe an hour or so.

**Comment by torkleyy** [comment]
So could we just allow switching the dispatcher for now, with a simple method the user has to call?

**Comment by Rhuagh** [comment]
Strictly speaking you can use resources to disable/enable systems right now

**Comment by torkleyy** [comment]
My concern here is not the time needed to implement it, but rather the way we're going with our overall design. They way we're going so fast with so little experience with how the low level stuff of Amethyst is used.

**Comment by torkleyy** [comment]
Yes you can, but I don't see a reason why not to allow using another dispatcher. Using a resource would add boilerplate and additional performance cost, switching the dispatcher is just adding a setter really - and overhead-free.

**Comment by Rhuagh** [comment]
My concern for the short term with that is renderer management, since that is local to the RenderSystem.

**Comment by jojolepro** [comment]
> Strictly speaking you can use resources to disable/enable systems right now

Sounds like global switches to me, and I'm pretty sure that using those will end up in a giant mess really quickly as soon as you get 2-3+ states  @Rhuagh 

Also I want this because it will make testing the new GUI easier.
I want to make a simple menu in my game, and when you click on the play button, it starts the game.
That means that the game system is added only after pressing the button. I also need to fire an event from the gui into the State on_event to change the state. @Xaeroxe Said that even could be handled by adding a 1 frame component "Clicked" on the button. I'm not happy with that, as I have to poll the components of each button each frame, instead of just firing an event (rip performance).

**Comment by omni-viral** [comment]
One can just have `Dispatcher` inside his `State` and run it every time `State::update` is called.

**Comment by Rhuagh** [comment]
Yes, global switches, which is doable right now as an _interim_ solution. Not a permanent solution.

**Comment by Rhuagh** [comment]
And in the topic of callbacks, that quickly leads to gigantic mess on multi threaded systems

**Comment by jojolepro** [comment]
That's why I talked about firing events that are catched in State::handle_event. Callbacks are terrible when used in multithreaded.

**Comment by Rhuagh** [comment]
When it comes to UI, and clicked component, either you or the UI system will need to do polling, so it's just a matter of where you prefer it to be.

**Comment by jojolepro** [comment]
We get an event when the mouse is clicked, so its at that moment that the UI system makes a single poll to know where the click was, and then it throws an event that the user has to deal with. Clicked<UIElement> or something.

**Comment by Xaeroxe** [comment]
@torkleyy I can prove this doesn't introduce any inflexibility, because I can implement it completely outside of the amethyst crate.

```rust
use specs::{Dispatcher, DispatcherBuilder, Entity};

struct MyState {
    dispatcher: Dispatcher;
    scene: Vec<Entity>;
}

impl State for MyState {
    fn on_start(&mut self, eng: &mut Engine) {
        self.dispatcher = DispatcherBuilder::new()
            // Build your dispatcher
            .build();
    }

    fn update(&mut self, eng: &mut Engine) -> Trans {
        self.dispatcher.dispatch(eng.world.res);
        None
    }

    fn on_stop(&mut self, eng: &mut Engine) {
        eng.world.delete_entities(&self.scene);
    }
}
```

So next question: If you can already do this downstream why introduce it into amethyst?  Answer: Because this architecture is highly useful for game development and not immediately obvious.

**Comment by OvermindDL1** [comment]
(I feel repetitive ^.^;)
In my engine I had multiple ECS 'worlds', each level would have one, the overall engine had one, even some entities in a level would have their own.  For example on a tile-based game I had a registration of primary tile types that was an ECS world on its own, this was a global state loaded via definition files at load time.  In a level another ECS referenced names in that outer ECS (the names being Atom64's of course, so a single 64-bit integer in size) in addition could have components on them for in-world tile data for specific locations (the tile names themselves were in a chunked array that was in a tree of chunks, each chunk being 256x256 in size, very fast lookups and basic rendering was based on these, so it was blazingly fast).  Such a world itself could be an entity in 'another' world (a ship was such a 'tile' world that flew around a larger world.  As long as I'm able to replicate all this efficiently, I'm good.  ^.^

**Comment by jojolepro** [comment]
So basically recursive universes/world? A universe in a universe in another universe.

**Comment by OvermindDL1** [comment]
> So basically recursive universes/world? A universe in a universe in another universe.

Effectively yeah, it was useful to build, say, a robot as tiles that did things on a ship that then flew around a larger universe.  It made it really trivial to both serialize and to keep very efficient (the worlds inside worlds only interacted at defined points, no entity crossover unless they were serialized out of one world and loaded into another) so it kept the entity count 'per-world' very low, which also helps with efficiency (only in a minor way in my engine in most cases, but in some it was quite valuable).

**Comment by Xaeroxe** [comment]
@OvermindDL1 Would said recursive worlds also require the global dispatcher to be executed on them?  Because that could get really complicated in a hurry.  You can also implement this completely downstream, **iff** you don't need the global dispatcher executed on your worlds.

**Comment by Xaeroxe** [comment]
@OvermindDL1 Would a hierarchy of entity transforms fulfill your needs?  Because we do have that.

**Comment by OvermindDL1** [comment]
> @OvermindDL1 Would said recursive worlds also require the global dispatcher to be executed on them? Because that could get really complicated in a hurry. You can also implement this completely downstream, iff you don't need the global dispatcher executed on your worlds.

Nope.  In fact each could even pause and then 'catch up' later when the player got near again.  They tended to be executed when a component on the owning entity was ran ('it' was the one that then ran the systems in its own world on the entity).

> @OvermindDL1 Would a hierarchy of entity transforms fulfill your needs? Because we do have that.

That is what I did anyway for the rendering side.  This distinction was more for the pure processing side, not the visual side.

However yes, it is easily possible to just put all entities in every single sub-object in the same world (with some kind of 'parent' component), but I ran into a lot of cases where performance became an issue then due to the sheer number and inability to individually pause groups of entities from process or to process them at different rates without adding components to hold that information and thread access of those components through near every single system (which is what my first incarnation of it did).

**Comment by Xaeroxe** [comment]
So yeah you can do that on Amethyst as is with a `World` Component that you just execute your own dispatchers on then.  If we find a significant need for that kind of processing down the line we could in the future also add a `World` component to Amethyst itself, but that's out of scope for this RFC right now.

**Comment by OvermindDL1** [comment]
> So yeah you can do that on Amethyst as is with a World Component that you just execute your own dispatchers on then. If we find a significant need for that kind of processing down the line we could in the future also add a World component to Amethyst itself, but that's out of scope for this RFC right now.

Perfect, just making sure before things get too much further along.  :-)

**Comment by jojolepro** [comment]
If the dispatcher is owner by the state, does that affect the % of multithreading that we can possibly achieve? @Xaeroxe

**Comment by Xaeroxe** [comment]
@jojolepro Absolutely, however there isn't any way around that.  Dispatchers can't have their system graphs modified after being created, so the best I can come up with is multiple dispatchers.  Each dispatcher can multithread independently, but they can't multithread with each other.

**Comment by jojolepro** [comment]
We are starting to have quite a few "stop the world" dispatcher, but I guess its better than no multithreading at all. There's one needed for the network entity sync, one for drawing(I think), one per state, and one global.
Offsubject question: Is the rendering being made independently of the changes to the data/ecs under it? Like it could draw always using the 2 most recent data that was copied to it, and do interpolation. Basically the ecs pushes the state to a buffer into the renderer, and the renderer figures out approximately what and where it should draw. (That's how I made it in my engine, I duped the data, but it allowed me to run the renderer at 1000 fps while the logic ran at 20 fps)

**Comment by Xaeroxe** [comment]
There is not a dispatcher for drawing at the moment, that is combined with the global dispatcher.  It might be possible to make some aspects of networking part of the global dispatcher too, but that'll be part of a much larger process.

**Comment by Xaeroxe** [comment]
I believe the current concerns with this RFC are resolved, so can I get a Yea or Nay from participants on this issue?  Specifically @jojolepro @torkleyy @omni-viral @Rhuagh and @OvermindDL1 .  If you're going to say Nay please express your concerns so they can be considered and possibly addressed.

**Comment by jojolepro** [comment]
I'd like to show you my design for the ecs-renderer decoupling design, as it can be reused at other places, and might affect this too. My phone has been uploading to imgur for the past 10 minutes though...

**Comment by Xaeroxe** [comment]
@jojolepro That seems concerned with engine internals, while this is concerned with the user facing side.  Perhaps we can continue that conversation elsewhere?

**Comment by jojolepro** [comment]
:+1: on this

**Comment by OvermindDL1** [comment]
I'm :+1: on this as well, definitely want the ability to temporarily pop a state when going to, say, the main menu to change options while in-game then restore it after (I'd traditionally use multiple ECS worlds for that as I'd never have the menu and game share entities anyway).

**Comment by Xaeroxe** [comment]
Oh uh to be clear @OvermindDL1 we already have the state machine described here, this is just proposing an enhancement to that state machine.

**Comment by Xaeroxe** [comment]
I just re-read your comment, you're talking about the Scene structure, my bad.

**Comment by Xaeroxe** [comment]
https://github.com/amethyst/amethyst/pull/380 This PR implements the idea in this RFC minus the inactivation of scene entities for states lower in the stack.  I'll implement that once I'm more confident in how I'd like to do that.

**Comment by Rhuagh** [comment]
So, the conclusion is that we don't integrate this into the engine core, we let the user do this "on their own", inside `State`?

**Comment by jojolepro** [comment]
:+1:

**Comment by Xaeroxe** [comment]
Since we've decided to use another plan I'm closing this issue.
